src/SFML/Window/Win32/JoystickImpl.cpp

```diff
@@ -44,1115 +44,724 @@
+        {
+            // that didn't work!
+            auto lastError = GetLastError();
+            err() << "HidP_GetCaps returned [" << result << "] unexpectedly! GetLastError: [" << lastError << "]"
+                  << std::endl;
+            return;
         }
 
-        // Buttons
-        for (unsigned int i = 0; i < Joystick::ButtonCount; ++i)
-            state.buttons[i] = (pos.dwButtons & (1u << i)) != 0;
+        joystickImpl.m_caps.buttonCount = hCaps.NumberInputButtonCaps;
+        auto axes                       = hCaps.NumberInputValueCaps;
+        for (std::size_t i = 0; i < Joystick::AxisCount && i < axes; ++i)
+            joystickImpl.m_caps.axes[getAxis(static_cast<int>(i))] = true;
     }
 
-    return state;
-}
-
-
-////////////////////////////////////////////////////////////
-void JoystickImpl::initializeDInput()
-{
-    // Try to load dinput8.dll
-    dinput8dll = LoadLibraryA("dinput8.dll");
-
-    if (dinput8dll)
+    unsigned int xInputIndex = 0;
+    for (auto& joystick : joysticks)
     {
-        // Try to get the address of the DirectInput8Create entry point
-        using DirectInput8CreateFunc = HRESULT(WINAPI*)(HINSTANCE, DWORD, const IID&, LPVOID*, LPUNKNOWN);
-        auto directInput8Create      = reinterpret_cast<DirectInput8CreateFunc>(
-            reinterpret_cast<void*>(GetProcAddress(dinput8dll, "DirectInput8Create")));
+        if (joystick.m_xInputIndex != 0xFFFFFFFF)
+            ++xInputIndex;
 
-        if (directInput8Create)
+        if (joystick.m_lastDeviceHandle == nullptr)
         {
-            // Try to acquire a DirectInput 8.x interface
-            const HRESULT result = directInput8Create(GetModuleHandleW(nullptr),
-                                                      0x0800,
-                                                      guids::IID_IDirectInput8W,
-                                                      reinterpret_cast<void**>(&directInput),
-                                                      nullptr);
-
-            if (FAILED(result))
-            {
-                // De-initialize everything
-                directInput = nullptr;
-                FreeLibrary(dinput8dll);
-                dinput8dll = nullptr;
+            // slot is "free"
+            if (joystickImpl.m_useXInput)
+                joystickImpl.m_xInputIndex = xInputIndex;
 
-                err() << "Failed to initialize DirectInput: " << result << std::endl;
-            }
-        }
-        else
-        {
-            // Unload dinput8.dll
-            FreeLibrary(dinput8dll);
-            dinput8dll = nullptr;
+            joystick = joystickImpl;
+            break;
         }
     }
-}
-
 
-////////////////////////////////////////////////////////////
-void JoystickImpl::cleanupDInput()
-{
-    // Release the DirectInput interface
-    if (directInput)
+    if (fileHandle)
     {
-        directInput->Release();
-        directInput = nullptr;
+        CloseHandle(fileHandle);
```

Comment:
I see a number of codepaths where the function exits without this being called. Perhaps we wrap `fileHandle` in a `std::unique_ptr` with a custom deleter so that this is guaranteed to be called when the function exits. 

See `sf::FileInputStream::m_file` and `sf::FileInputStream::FileCloser` as a good example of this pattern.

Reply:
On second thought this is a bit tricker since `HANDLE` is a `void*` and you can't have `std::unique_ptr<void>`. I suppose we could use `std::shared_ptr<void>` but that's less appealing. It may still be necessary though given the multiple exit paths of this function.